execution/shuffle.go

```diff
@@ -0,0 +1,449 @@
+
+	for partition, node := range s.sources {
+		sourceStreamID, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString(fmt.Sprintf("shuffle_input_%d", partition)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream ID for shuffle with ID %s", shuffleID.Id)
+		}
+
+		newPipelineMetadata := PipelineMetadata{
+			NextShuffleID: nextShuffleID,
+			Partition:     partition,
+		}
+
+		// Get the shuffle source stream.
+		rs, execOutput, err := node.Get(context.WithValue(ctx, pipelineMetadataContextKey{}, newPipelineMetadata), variables, sourceStreamID)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream for shuffle with ID %s", shuffleID.Id)
+		}
+
+		// Add deduplicated next shuffle information.
+		for id, data := range execOutput.NextShuffles {
+			nextShuffles[id] = data
+		}
+
+		senderStreamID, err := GetSourceStreamID(tx.WithPrefix(shuffleID.AsPrefix()), octosql.MakeString(fmt.Sprintf("shuffle_sender_%d", partition)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get new source stream ID for shuffle with ID %s", shuffleID.Id)
+		}
+
+		strategy, err := s.strategyPrototype.Get(ctx, variables)
+		if err != nil {
+			return nil, errors.Wrapf(err, "couldn't get strategy for shuffle with ID %s", shuffleID.Id)
+		}
+
+		// Start the shuffle sender.
+		sender := NewShuffleSender(senderStreamID, shuffleID, strategy, s.outputPartitionCount, partition)
+
+		engine := NewPullEngine(sender, stateStorage, rs, nil, execOutput.WatermarkSource, false)
+
+		go engine.Run(ctx)
+	}
+
+	return nextShuffles, nil
+}
+
+type ShuffleReceiver struct {
+	streamID             *StreamID
+	shuffleID            *ShuffleID
+	sourcePartitionCount int
+	partition            int
+
+	received int
+}
+
+func NewShuffleReceiver(streamID *StreamID, shuffleID *ShuffleID, sourcePartitionCount int, partition int) *ShuffleReceiver {
+	return &ShuffleReceiver{
+		streamID:             streamID,
+		shuffleID:            shuffleID,
+		sourcePartitionCount: sourcePartitionCount,
+		partition:            partition,
+	}
+}
+
+var watermarksPrefix = []byte("$watermarks$")
+
+func getQueuePrefix(from, to int) []byte {
+	return []byte(fmt.Sprintf("$to_%d$from_%d$", to, from))
+}
+
+var endsOfStreamsPrefix = []byte("$ends_of_streams$")
+
+func (rs *ShuffleReceiver) Next(ctx context.Context) (*Record, error) {
+	tx := storage.GetStateTransactionFromContext(ctx)
+	streamPrefixedTx := tx.WithPrefix(rs.streamID.AsPrefix())
+	endOfStreamsMap := storage.NewMap(streamPrefixedTx.WithPrefix(endsOfStreamsPrefix))
+	watermarksMap := storage.NewMap(streamPrefixedTx.WithPrefix(watermarksPrefix))
+
+	// We want to randomize the order so we don't always read from the first input if records are available.
+	sourceOrder := rand.Perm(rs.sourcePartitionCount)
+
+	changeSubscriptions := make([]*storage.Subscription, rs.sourcePartitionCount)
+
+sourcePartitionsLoop:
```

Comment:
Just to make sure I understand what's happening here: first you try to get a record from a source. If you receive a WaitForChanges you save it, because all sources might make you wait, if you get an actual record you cancel all the subscriptions you have previously made and just return the record you got. Is that correct?

Reply:
Yeah